### PR TITLE
Upgrade to Python 3.12.7

### PR DESCRIPTION
**Pull request by cclauss**
### Description

As discussed [here](https://github.com/pyodide/pyodide/pull/3741#issuecomment-2308959591), let's try upgrading from Python 3.12.1 to 3.12.7.

https://github.com/pyodide/pyodide/blob/main/docs/development/maintainers.md#upgrading-pyodide-to-a-new-version-of-cpython
https://github.com/pyodide/pyodide/blob/main/docs/development/maintainers.md#old-major-python-upgrades
```diff
-    - image: pyodide/pyodide-env:20240928-chrome127-firefox128
+    - image: pyodide/pyodide-env:20241106-chrome130-firefox132
```

Issues:
* [x] The Docker image has an OLDER date than the image replaces but has newer versions of the browsers.
    * Should this PR start with a more current Docker image?
* [x] Python 3.12.7 has already been released.
    * ~Should this PR focus on 3.12.6 or~ 3.12.7? 
* [x] Two `test-python` pytests related to positional-only parameters fail:
    * `test_pyodide.py::relaxed_call`
    * `test_pyodide.py::relaxed_wrap`
* [x] The `test-core (macos-14, selenium, safari)` tests fail.
    * SeleniumHQ/selenium#14698
    * SeleniumHQ/selenium#14699
    * Also mentioned in #5152
    * Fixed but not yet released.  https://github.com/SeleniumHQ/selenium/releases
* [x] `sphinx-js 3.2.1 requires markupsafe==2.0.1, but you have markupsafe 3.0.2 which is incompatible.`
    * https://github.com/pyodide/pyodide/blob/0adc6d050ac2513396cdd0eade2bd4b01849370d/docs/requirements-doc.txt#L25  Can we unpin `sphinx-js`?  Unsorted `requirements` files are painful to read.
    * mozilla/sphinx-js#243

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry

@okken

### CODE REVIEW

**Comment by hoodmane** [comment]
I can generate another docker image. But indeed, should it be 3.12.6 or 3.12.7?

**Comment by hoodmane** [comment]
The test-python tests are failing because this PR fixed the error messages raised by `bind()`. Nice to see that someone is fixing these.
https://github.com/python/cpython/pull/103404/

**Comment by cclauss** [comment]
3.12.7 please

**Comment by hoodmane** [comment]
I'm currently pushing pyodide/pyodide-env:20241106-chrome130-firefox132, should be available in a few minutes.

**Comment by ryanking13** [comment]
> The docker image won't build because I'm getting a 404 from
[https://dl.google.com/linux/chrome/deb/pool/main/g/google-chrome-stable/google-chrome-stable_${CHROME_VERSION_FULL}-1_amd64.deb](https://dl.google.com/linux/chrome/deb/pool/main/g/google-chrome-stable/google-chrome-stable_$%7BCHROME_VERSION_FULL%7D-1_amd64.deb)
@ryanking13?

Hmm, does the variable not replaced? I cannot reproduce it locally. Anyway, we'll probably add some automated test to test our dockerfile.

**Comment by cclauss** [comment]
Is this PR blocked by
* #5165?

**Comment by hoodmane** [comment]
So it looks like for the core tests we need to skip or xfail `leakers.test_ctypes` and we need to update the expected error message in `test_array_slice_assign_2`.

File: src/tests/test_jsproxy.py
Comment (hoodmane) [inline src/tests/test_jsproxy.py:1524]: I guess we want to update the code so that the error that the JS Array raises matches the new Python error message.

Comment (cclauss) [reply src/tests/test_jsproxy.py]: I tried modifying `src/core/jsproxy.c` but cannot figure out how to modify `exc_info_3a.value.args` without also modifying `exc_info_3b.value.args`.

https://github.com/search?q=repo%3Apyodide%2Fpyodide+%22can+only+assign+an+iterable%22&type=code

Comment (hoodmane) [reply src/tests/test_jsproxy.py]: I'm not sure what you mean. The test passes if I apply this change:`
```patch
diff --git a/src/core/jsproxy.c b/src/core/jsproxy.c
index 711b15e1..4e3b3d09 100644
--- a/src/core/jsproxy.c
+++ b/src/core/jsproxy.c
@@ -1698,7 +1698,7 @@ JsArray_ass_subscript(PyObject* self, PyObject* item, PyObject* pyvalue)
     slicelength = PySlice_AdjustIndices(length, &start, &stop, step);
 
     if (pyvalue != NULL) {
-      seq = PySequence_Fast(pyvalue, "can only assign an iterable");
+      seq = PySequence_Fast(pyvalue, "must assign iterable to extended slice");
       FAIL_IF_NULL(seq);
     }
     if (pyvalue != NULL && step != 1 &&
diff --git a/src/tests/test_jsproxy.py b/src/tests/test_jsproxy.py
index d8208ecc..26bc88a5 100644
--- a/src/tests/test_jsproxy.py
+++ b/src/tests/test_jsproxy.py
@@ -1520,8 +1520,7 @@ def test_array_slice_assign_2(selenium):
 
     assert exc_info_1a.value.args == exc_info_1b.value.args
     assert exc_info_2a.value.args == exc_info_2b.value.args
-    assert exc_info_3a.value.args == ("must assign iterable to extended slice",)
-    assert exc_info_3b.value.args == ("can only assign an iterable",)
+    assert exc_info_3a.value.args == exc_info_3b.value.args
 
 
 @std_hypothesis_settings
```

**Comment by cclauss** [comment]
How could so much change between Python 3.12.1 and 3.12.7?
How could testing the abstract_base_classes module fail?
How can I run these tests on localhost?
Is there some other way of quickly discovering the failing tests?
Should we be waiting for a release that fixes SeleniumHQ/selenium#14699 or is the downgraded Selenium OK to use?
I tried running `python src/tests/make_test_list.py` locally but that resulted in an empty list.

**Comment by hoodmane** [comment]
I'll look into failing tests. You should be able to run them locally with:
```
python3.12 -m venv .venv
source .venv/bin/activate
pip install -r requirements.txt
make
pytest src/tests/test_core_python.py
```

**Comment by hoodmane** [comment]
Are you getting:
```
Dynamic linking error: cannot resolve symbol _PyTestCapi_Init_Run
```
it's a bit perplexing I'll look into it.

**Comment by hoodmane** [comment]
Opened a PR to fix `make_test_list.py` :
https://github.com/pyodide/pyodide/pull/5177

**Comment by cclauss** [comment]
It was the `make` step that I missed...  Kinda important!  Thx.

**Comment by hoodmane** [comment]
`latest` is perhaps misnamed, it's actually unstable tip of tree. I imagine we'll just release this as part of Pyodide 0.27.0.

**Comment by hoodmane** [comment]
Okay core test suite looking good. @ryanking13 can you look at the pygame-ce build error?

**Comment by cclauss** [comment]
`ci/circleci: build-packages-no-numpy-dependents` — Tests failed on CircleCI on `pygame-ce`.

https://github.com/pyodide/pyodide/blob/912ccca6f0fc9722e2e69ec2ad1868d62c27c10a/packages/pygame-ce/meta.yaml#L7
vs. pygame-ce 2.5.2 at https://pypi.org/project/pygame-ce however this repo contains a few patches for `pygame-ce`.

**Comment by ryanking13** [comment]
> I'm okay with using the downgraded selenium for now. @ryanking13 you have any objection?

No objection.

**Comment by ryanking13** [comment]
> Okay core test suite looking good. @ryanking13 can you look at the pygame-ce build error?

**Comment by ryanking13** [comment]
It seems like pygame-ce 2.5.0 introduced meson build system which is not yet compatible to Emscripten, for now supressing the error message with `-Wno-unreachable-code-fallthrough` seems to work. See: https://github.com/ryanking13/pyodide/commit/bc668a7375e8ca1f6516996620e05e2e6499b55b#diff-8528a6d7acaacb6cc8a9be005a6cc0b6b978bdd0a20c061c8a12866fc1de29b5R34

**Comment by cclauss** [comment]
### `benchmark-stack-size`
% `pytest -s benchmark/stack_usage.py --rt node,chrome,firefox | sed -n 's/## //pg'`
```
browser  py_limit  py_usage  js_depth  py_depth  js_depth/py_usage  js_depth/py_depth
chrome       1000     38.10     12528      1997                328                  6
node         1000     54.89     13665      1997                248                  6

Exited with code exit status 1
```
### No firefox!

**Comment by hoodmane** [comment]
I think `test_fstring-chrome` is a flake, I'll try rerunning it. It took 21.78 seconds when I ran it locally so it could be a timeout. Anyways, it surely works fine in practice so if need be we can xfail it.

**Comment by cclauss** [comment]
Any hints on the [benchmark-stack-size](https://github.com/pyodide/pyodide/pull/5149#issuecomment-2466111434) failure before I rebase and xfail `test_fstring-chrome`?

**Comment by hoodmane** [comment]
Well I don't reproduce the firefox benchmark-stack-size failure locally

**Comment by hoodmane** [comment]
With rerun in ssh I can see that it is failing in CI. I don't really understand why this PR would break it. Maybe let's just turn off the firefox stack size benchmark for now...

**Comment by hoodmane** [comment]
> sphinx-js 3.2.1 requires markupsafe==2.0.1, but you have markupsafe 3.0.2 which is incompatible.

We use a fork of sphinx-js:
https://github.com/pyodide/sphinx-js-fork/blob/master/setup.py#L22C11-L22C55

You can make a PR there and then update the commit we use.

**Comment by cclauss** [comment]
* pyodide/sphinx-js-fork#170

**Comment by cclauss** [comment]
Arrgghh…. https://github.com/robotraconteur/robotraconteur ?!?

**Comment by cclauss** [comment]
Please rerun the failing test to see if this is a flake?

**Comment by hoodmane** [comment]
Retriggered CI.

**Comment by hoodmane** [comment]
Can you add a changelog entry? Other than that looks good to go if CI passes.

**Comment by cclauss** [comment]
Dedicated to the great work of @okken

As he implied at https://www.youtube.com/watch?v=EB2emah5d5g&t=9m40s Python 3.12.1 is a bit out of date ;-)

**Comment by hoodmane** [comment]
We try to stay on a version that is at most 18 months old =)

**Comment by hoodmane** [comment]
Thanks @cclauss!

**Review by hoodmane — APPROVED** [review]

**Comment by cclauss** [comment]
### Was and update to `pyodide-cross-build-environments.json` required?!?

**Comment by hoodmane** [comment]
No, only after Pyodide is released.

**Comment by cclauss** [comment]
Great!  https://pyodide.org/en/latest/console.html is now 3.12.7.

**Pull request merged**
